File: src/test/function_test/test_incr.cpp
// Copyright (c) 2017, Xiaomi, Inc.  All rights reserved.
// This source code is licensed under the Apache License Version 2.0, which
// can be found in the LICENSE file in the root directory of this source tree.

#include <cstdlib>
#include <string>
#include <vector>
#include <climits>
#include <map>

#include <dsn/service_api_c.h>
#include <unistd.h>
#include <pegasus/client.h>
#include <gtest/gtest.h>

using namespace ::pegasus;

extern pegasus_client *client;

TEST(incr, unexist_key)
{
    int ret = client->del("incr_test_unexist_key", "");
    ASSERT_EQ(0, ret);

    int64_t new_value_int;
    ret = client->incr("incr_test_unexist_key", "", 100, new_value_int);
    ASSERT_EQ(0, ret);
    ASSERT_EQ(100, new_value_int);

    std::string new_value_str;
    ret = client->get("incr_test_unexist_key", "", new_value_str);
    ASSERT_EQ(0, ret);
    ASSERT_EQ("100", new_value_str);

    ret = client->del("incr_test_unexist_key", "");
    ASSERT_EQ(0, ret);
}

TEST(incr, empty_key)
{
    int ret = client->set("incr_test_empty_key", "", "");
    ASSERT_EQ(0, ret);

    int64_t new_value_int;
    ret = client->incr("incr_test_empty_key", "", 100, new_value_int);
    ASSERT_EQ(0, ret);
    ASSERT_EQ(100, new_value_int);

    std::string new_value_str;
    ret = client->get("incr_test_empty_key", "", new_value_str);
    ASSERT_EQ(0, ret);
    ASSERT_EQ("100", new_value_str);

    ret = client->del("incr_test_empty_key", "");
    ASSERT_EQ(0, ret);
}

TEST(incr, negative_value)
{
    int ret = client->set("incr_test_negative_value", "", "-100");
    ASSERT_EQ(0, ret);

    int64_t new_value_int;
    ret = client->incr("incr_test_negative_value", "", -1, new_value_int);
    ASSERT_EQ(0, ret);
    ASSERT_EQ(-101, new_value_int);

    std::string new_value_str;
    ret = client->get("incr_test_negative_value", "", new_value_str);
    ASSERT_EQ(0, ret);
    ASSERT_EQ("-101", new_value_str);

    ret = client->del("incr_test_negative_value", "");
    ASSERT_EQ(0, ret);
}

TEST(incr, increase_zero)
{
    int ret = client->set("incr_test_increase_zero", "", "100");
    ASSERT_EQ(0, ret);

    int64_t new_value_int;
    ret = client->incr("incr_test_increase_zero", "", 0, new_value_int);
    ASSERT_EQ(0, ret);
    ASSERT_EQ(100, new_value_int);

    std::string new_value_str;
    ret = client->get("incr_test_increase_zero", "", new_value_str);
    ASSERT_EQ(0, ret);
    ASSERT_EQ("100", new_value_str);

    ret = client->del("incr_test_increase_zero", "");
    ASSERT_EQ(0, ret);
}

TEST(incr, multiple_increment)
{
    int ret = client->set("incr_test_multiple_increment", "", "100");
    ASSERT_EQ(0, ret);

    int64_t new_value_int;
    ret = client->incr("incr_test_multiple_increment", "", 1, new_value_int);
    ASSERT_EQ(0, ret);
    ASSERT_EQ(101, new_value_int);

    std::string new_value_str;
    ret = client->get("incr_test_multiple_increment", "", new_value_str);
    ASSERT_EQ(0, ret);
    ASSERT_EQ("101", new_value_str);

    ret = client->incr("incr_test_multiple_increment", "", 2, new_value_int);
    ASSERT_EQ(0, ret);
    ASSERT_EQ(103, new_value_int);

    ret = client->get("incr_test_multiple_increment", "", new_value_str);
    ASSERT_EQ(0, ret);
    ASSERT_EQ("103", new_value_str);

    ret = client->del("incr_test_multiple_increment", "");
    ASSERT_EQ(0, ret);
}

TEST(incr, invalid_old_data)
{
    int ret = client->set("incr_test_invalid_old_data", "", "aaa");
    ASSERT_EQ(0, ret);

    int64_t new_value_int;
    ret = client->incr("incr_test_invalid_old_data", "", 1, new_value_int);
    ASSERT_EQ(PERR_INVALID_ARGUMENT, ret);
    ASSERT_EQ(0, new_value_int);

    std::string new_value_str;
    ret = client->get("incr_test_invalid_old_data", "", new_value_str);
    ASSERT_EQ(0, ret);
    ASSERT_EQ("aaa", new_value_str);

    ret = client->del("incr_test_invalid_old_data", "");
    ASSERT_EQ(0, ret);
}

TEST(incr, out_of_range_old_data)
{
    int ret = client->set(
        "incr_test_out_of_range_old_data", "", "10000000000000000000000000000000000000");
    ASSERT_EQ(0, ret);

    int64_t new_value_int;
    ret = client->incr("incr_test_out_of_range_old_data", "", 1, new_value_int);
    ASSERT_EQ(PERR_INVALID_ARGUMENT, ret);
    ASSERT_EQ(0, new_value_int);

    std::string new_value_str;
    ret = client->get("incr_test_out_of_range_old_data", "", new_value_str);
    ASSERT_EQ(0, ret);
    ASSERT_EQ("10000000000000000000000000000000000000", new_value_str);

    ret = client->del("incr_test_out_of_range_old_data", "");
    ASSERT_EQ(0, ret);
}

TEST(incr, up_overflow)
{
    int ret = client->set("incr_test_up_overflow", "", "1");
    ASSERT_EQ(0, ret);

    int64_t new_value_int;
    ret = client->incr("incr_test_up_overflow", "", LLONG_MAX, new_value_int);
    ASSERT_EQ(PERR_INVALID_ARGUMENT, ret);
    ASSERT_EQ(1, new_value_int);

    std::string new_value_str;
    ret = client->get("incr_test_up_overflow", "", new_value_str);
    ASSERT_EQ(0, ret);
    ASSERT_EQ("1", new_value_str);

    ret = client->del("incr_test_up_overflow", "");
    ASSERT_EQ(0, ret);
}

TEST(incr, down_overflow)
{
    int ret = client->set("incr_test_down_overflow", "", "-1");
    ASSERT_EQ(0, ret);

    int64_t new_value_int;
    ret = client->incr("incr_test_down_overflow", "", LLONG_MIN, new_value_int);
    ASSERT_EQ(PERR_INVALID_ARGUMENT, ret);
    ASSERT_EQ(-1, new_value_int);

    std::string new_value_str;
    ret = client->get("incr_test_down_overflow", "", new_value_str);
    ASSERT_EQ(0, ret);
    ASSERT_EQ("-1", new_value_str);

    ret = client->del("incr_test_down_overflow", "");
    ASSERT_EQ(0, ret);
}

TEST(incr, preserve_ttl)
{
    int ret = client->set("incr_test_preserve_ttl", "", "100", 5000, 3);
    ASSERT_EQ(0, ret);

    int64_t new_value_int;
    ret = client->incr("incr_test_preserve_ttl", "", 1, new_value_int);
    ASSERT_EQ(0, ret);
    ASSERT_EQ(101, new_value_int);

    int ttl_seconds;
    ret = client->ttl("incr_test_preserve_ttl", "", ttl_seconds);
    ASSERT_EQ(0, ret);
    ASSERT_GE(3, ttl_seconds);

    sleep(4);

    std::string new_value_str;
    ret = client->get("incr_test_preserve_ttl", "", new_value_str);
    ASSERT_EQ(PERR_NOT_FOUND, ret);

    ret = client->del("incr_test_preserve_ttl", "");
    ASSERT_EQ(0, ret);
}

TEST(incr, reset_ttl)
{
    /// reset after old value ttl timeout
    int ret = client->set("incr_test_reset_ttl", "", "100", 5000, 3);
    ASSERT_EQ(0, ret);

    sleep(4);

    int64_t new_value_int;
    ret = client->incr("incr_test_reset_ttl", "", 1, new_value_int);
    ASSERT_EQ(0, ret);
    ASSERT_EQ(1, new_value_int);

    int ttl_seconds;
    ret = client->ttl("incr_test_reset_ttl", "", ttl_seconds);
    ASSERT_EQ(0, ret);
    ASSERT_GE(-1, ttl_seconds);

    ret = client->del("incr_test_reset_ttl", "");
    ASSERT_EQ(0, ret);

    /// reset with new ttl
    ret = client->set("incr_test_reset_ttl", "", "100");
    ASSERT_EQ(0, ret);

    ret = client->ttl("incr_test_reset_ttl", "", ttl_seconds);
    ASSERT_EQ(0, ret);
    ASSERT_GE(-1, ttl_seconds);

    ret = client->incr("incr_test_reset_ttl", "", 1, new_value_int, 5000, 10);
    ASSERT_EQ(0, ret);
    ASSERT_EQ(101, new_value_int);

    ret = client->ttl("incr_test_reset_ttl", "", ttl_seconds);
    ASSERT_EQ(0, ret);
    ASSERT_LT(0, ttl_seconds);
    ASSERT_GE(10, ttl_seconds);

    ret = client->del("incr_test_reset_ttl", "");
    ASSERT_EQ(0, ret);

    /// reset with no ttl
    ret = client->set("incr_test_reset_ttl", "", "200", 5000, 10);
    ASSERT_EQ(0, ret);

    ret = client->ttl("incr_test_reset_ttl", "", ttl_seconds);
    ASSERT_EQ(0, ret);
    ASSERT_LT(0, ttl_seconds);
    ASSERT_GE(10, ttl_seconds);

    ret = client->incr("incr_test_reset_ttl", "", 1, new_value_int, 5000, -1);
    ASSERT_EQ(0, ret);
    ASSERT_EQ(201, new_value_int);

    ret = client->ttl("incr_test_reset_ttl", "", ttl_seconds);
    ASSERT_EQ(0, ret);
    ASSERT_GE(-1, ttl_seconds);

    ret = client->del("incr_test_reset_ttl", "");
    ASSERT_EQ(0, ret);
}

File: java-client/src/test/java/com/xiaomi/infra/pegasus/client/TestIncr.java
// Copyright (c) 2017, Xiaomi, Inc.  All rights reserved.
// This source code is licensed under the Apache License Version 2.0, which
// can be found in the LICENSE file in the root directory of this source tree.
package com.xiaomi.infra.pegasus.client;

/** @author qinzuoyan */
import org.junit.Assert;
import org.junit.Test;

/** Created by mi on 18-7-10. */
public class TestIncr {
  @Test
  public void testIncr() throws PException {
    PegasusClientInterface client = PegasusClientFactory.getSingletonClient();
    String tableName = "temp";

    byte[] hashKey = "incrkeyforjava".getBytes();
    byte[] sortKey = "0".getBytes();
    byte[] value = "".getBytes();
    int ttlSeconds = 0;

    try {
      System.out.println("set value ...");
      client.set(tableName, hashKey, sortKey, value, 0);
      System.out.println("set value ok");

      System.out.println("incr to empty value ...");
      long result = client.incr(tableName, hashKey, sortKey, 100);
      Assert.assertEquals(100, result);
      ttlSeconds = client.ttl(tableName, hashKey, sortKey);
      Assert.assertEquals(-1, ttlSeconds);
      System.out.println("incr to empty value ok");

      System.out.println("incr zero ...");
      result = client.incr(tableName, hashKey, sortKey, 0);
      Assert.assertEquals(100, result);
      ttlSeconds = client.ttl(tableName, hashKey, sortKey);
      Assert.assertEquals(-1, ttlSeconds);
      System.out.println("incr zero ok");

      System.out.println("incr negative ...");
      result = client.incr(tableName, hashKey, sortKey, -1);
      Assert.assertEquals(99, result);
      ttlSeconds = client.ttl(tableName, hashKey, sortKey);
      Assert.assertEquals(-1, ttlSeconds);
      System.out.println("incr negative ok");

      System.out.println("del value ...");
      client.del(tableName, hashKey, sortKey);
      System.out.println("del value ok");

      System.out.println("incr to un-exist value ...");
      result = client.incr(tableName, hashKey, sortKey, 200);
      Assert.assertEquals(200, result);
      ttlSeconds = client.ttl(tableName, hashKey, sortKey);
      Assert.assertEquals(-1, ttlSeconds);
      System.out.println("incr to un-exist value ok");

      System.out.println("incr with ttlSeconds > 0 ...");
      result = client.incr(tableName, hashKey, sortKey, 1, 10);
      Assert.assertEquals(201, result);
      ttlSeconds = client.ttl(tableName, hashKey, sortKey);
      Assert.assertTrue(ttlSeconds > 0);
      Assert.assertTrue(ttlSeconds <= 10);
      System.out.println("incr with ttlSeconds > 0 ok");

      System.out.println("incr with ttlSeconds == 0 ...");
      result = client.incr(tableName, hashKey, sortKey, 1);
      Assert.assertEquals(202, result);
      ttlSeconds = client.ttl(tableName, hashKey, sortKey);
      Assert.assertTrue(ttlSeconds > 0);
      Assert.assertTrue(ttlSeconds <= 10);
      System.out.println("incr with ttlSeconds == 0 ok");

      System.out.println("incr with ttlSeconds > 0 ...");
      result = client.incr(tableName, hashKey, sortKey, 1, 20);
      Assert.assertEquals(203, result);
      ttlSeconds = client.ttl(tableName, hashKey, sortKey);
      Assert.assertTrue(ttlSeconds > 10);
      Assert.assertTrue(ttlSeconds <= 20);
      System.out.println("incr with ttlSeconds > 0 ok");

      System.out.println("incr with ttlSeconds == -1 ...");
      result = client.incr(tableName, hashKey, sortKey, 1, -1);
      Assert.assertEquals(204, result);
      ttlSeconds = client.ttl(tableName, hashKey, sortKey);
      Assert.assertEquals(-1, ttlSeconds);
      System.out.println("incr with ttlSeconds == -1 ok");

      System.out.println("del value ...");
      client.del(tableName, hashKey, sortKey);
      System.out.println("del value ok");
    } catch (PException e) {
      e.printStackTrace();
      Assert.assertTrue(false);
    }

    PegasusClientFactory.closeSingletonClient();
  }
}
